[v2] Adding prometheus metrics for when the HPA using external metrics.

tbickford opened the pull request:
Adding a metrics server that produces Prometheus metrics (an error count and metric measure/value) for the external metrics used by only the HPA. Examples:

```
# HELP keda_hpa_scaler_errors_total Number of scaler errors
# TYPE keda_hpa_scaler_errors_total counter
keda_hpa_scaler_errors_total{metric="queuelength",namespace="default",selector="deploymentName=rabbitmq-keda-consumer-test"} 0
# HELP keda_hpa_scaler_metrics_value Metric Value
# TYPE keda_hpa_scaler_metrics_value gauge
keda_hpa_scaler_metrics_value{metric="queueLength",namespace="default",selector="deploymentName=rabbitmq-keda-consumer-test"} 17
```

Relates to issue #823 

I am open to suggestions for what tests should be added.

I am wondering if the initialization of the metrics server should be placed in the adapter main.go and make the address and path of the metrics configurable.  Thoughts?

zroubalik commented:
@tbickford this is great. I think that it would be great if you can target this for v2, there are already exposed  enpoints such as `/healtz` for liveness and readiness probes.

zroubalik commented:
> I am wondering if the initialization of the metrics server should be placed in the adapter main.go and make the address and path of the metrics configurable. Thoughts?

+1

tbickford commented:
Thanks @zroubalik! I'll rebase off of the `v2` branch. Good point on the health endpoints - I'll remove. I'll also work on making the address and path configurable.

zroubalik commented on cmd/adapter/main.go:
```diff
@@ -33,6 +33,8 @@ type Adapter struct {
 
 var logger = klogr.New().WithName("keda_metrics_adapter")
+var metricsPort int
+var metricsServer prommetrics.Server
```
Minor nit. Might be better to avoid naming these vars `metricsServer` and `metricsPort`. This whole adapter is a metrics server (just a different kind) so to avoid confusion I'd rename it, eg. `prometheusServer` or `promMetricsServer`, etc.

tbickford replied:
Totally agree. Good suggestion. I'll revise.

zroubalik commented on pkg/metrics/prometheus_metrics.go:
```diff
@@ -0,0 +1,70 @@
+	return &prometheusMetricServer{registry: registry}
+}
+
+// StartServer starts an HTTP serve to serve metrics
```
```suggestion
// StartServer starts an HTTP server to serve metrics
```

zroubalik commented on pkg/provider/provider.go:
```diff
@@ -25,6 +26,7 @@ type KedaProvider struct {
 	externalMetrics  []externalMetric
 	scaleHandler     scaling.ScaleHandler
 	watchedNamespace string
+	metricServer     prommetrics.Server
```
the same remark on naming as in adapter.go

zroubalik commented on pkg/provider/provider.go:
```diff
@@ -93,7 +96,12 @@ func (p *KedaProvider) GetExternalMetric(namespace string, metricSelector labels
+						p.metricServer.RecordHPAScalerMetrics(namespace, metric.MetricName, metricSelector.String(), metricValue)
+					}
 					matchingMetrics = append(matchingMetrics, metrics...)
+					p.metricServer.RecordHPAScalerErrorTotals(namespace, info.Metric, metricSelector.String(), err)
```
Shouldn't this be in the `if err != nil {}` block above?

tbickford replied:
I am glad you mentioned this. The reasoning behind it's placement is to handle the initial registration of the error metric - https://github.com/tbickford/keda/blob/edc011bf19378ae39e6269d0a336654365ad9083/pkg/metrics/prometheus_metrics.go#L65 (example - `keda_hpa_scaler_errors_total{metric="queuelength",namespace="default",selector="deploymentName=rabbitmq-keda-consumer-test"} 0`). I am not super happy with this logic. Alternatively I could do try to initially register the metric for each Scaler on startup but then I would also have to handle newly registered ScaledObjects which may not really save much from the performance and coding perspective. What are your thoughts?

zroubalik replied:
OK, but still, I might be missing something, but where do you actually record the erros? Currently the `p.metricsExporter.RecordHPAScalerErrorTotals(namespace, info.Metric, metricSelector.String(), err)` call is in the block where `err` is always `nil`

tbickford replied:
Nice catch - that is a mistake. It should be outside the block - I will correct.

tbickford commented:
Marking this for review - I noticed it was beginning to fall behind in with the rest of commit history

ahmelsayed commented on pkg/metrics/prometheus_metrics.go:
```diff
@@ -0,0 +1,84 @@
+// NewPrometheusMetricsExporter return a new instance of the prometheus metrics server
+func NewPrometheusMetricsExporter() Exporter {
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(scalerErrors)
+	registry.MustRegister(scalerMetricsValue)
+	registry.MustRegister(errorTotals)
```
I'm not really familiar with these prometheus apis, but from the usage of those vars (`scalerErrors`, `scalerMetricValue`, `errorTotals`) below, it seems that they are stateful per-exporter, right?

It's not a big deal since `NewPrometheusMetricsExporter` is only called once, but I think the contract of a New function usually would not assume shared state between returned values.

tbickford replied:
Thank you @ahmelsayed. The vars represent the metrics to be tracked (registered) with the Prometheus client registry which is used to expose the metrics for collection (the http endpoint). I suppose it would be better to move the registration of the metrics to an init function. I could also move the registry out of the factory method and just reference it directly when starting the Prometheus client handler since it is sharing state.

tomkerkhove commented:
Is `keda_hpa_scaler_errors_total` specific to a scaled object or a scaler in general?

This PR looks good to me but is solely focussed on HPA related metrics while we could surface more KEDA-oriented metrics as well I think.

I was wondering if it would make sense to introduce:
- A metric for amount of scaledobjects that are deployed and how many are in error?
  - `keda_hpa_scaler_errors_total` sounds scaler related and don't surface the impact on all the scaled objects
  - Could have a label with the trigger type to understand what people are mainly scaling on inside the cluster

But this can be a dedicated issue if you want; but see value in this as well.

Let me know if there are docs to review!

tbickford commented:
`keda_hpa_scaler_errors_total` is the total count error for all scalers. The total scaled objects and errors along with the triggers would also be tracked at least by namespace and as often as the `GetExternalMetric` function is being called. The big thing to note is these metrics are solely for the metrics adapter so when replica count > 1 (HPA managed scaling).  I'll work on adding some of the suggestions in the PR. I have not added anything in the docs yet but certainly willing to do so. Is there a particular spot this should go?

tomkerkhove commented:
> `keda_hpa_scaler_errors_total` is the total count error for all scalers. The total scaled objects and errors along with the triggers would also be tracked at least by namespace and as often as the `GetExternalMetric` function is being called.

That's ok, in that case a label for the scaler name would maybe be interesting to add?


> The big thing to note is these metrics are solely for the metrics adapter so when replica count > 1 (HPA managed scaling).

That's ok for me!

> I have not added anything in the docs yet but certainly willing to do so. Is there a particular spot this should go?

I'm thinking of a new section called "Operate" here:

![image](https://user-images.githubusercontent.com/4345663/89036976-02679f80-d33e-11ea-8847-33f4a5c88cd1.png)

What do you think?

tbickford commented:
I'll add in the scaler name as well. Sounds great.

nisan270390 commented on pkg/provider/provider.go:
```diff
@@ -93,8 +96,13 @@ func (p *KedaProvider) GetExternalMetric(namespace string, metricSelector labels
+					}
 					matchingMetrics = append(matchingMetrics, metrics...)
 				}
+				p.metricsExporter.RecordHPAScalerErrors(namespace, info.Metric, metricSelector.String(), err)
```
It seems that if errors returned before scales loop then keda_hpa_scaler_error_totals does not increased (like in this [line](https://github.com/tbickford/keda/blob/3974023c90a477518ee1e6451e0357cb884ecbf8/pkg/provider/provider.go#L86)).
Is it supposed to be like this?

tbickford replied:
That is a good point. The prometheus metric as it currently is expects the metric name as a label. If an error occurs in retrieving the scaler then there would be no metric so it may make sense to track errors by scaled object as well as by scaler.

tbickford commented:
> > `keda_hpa_scaler_errors_total` is the total count error for all scalers. The total scaled objects and errors along with the triggers would also be tracked at least by namespace and as often as the `GetExternalMetric` function is being called.
> 
> That's ok, in that case a label for the scaler name would maybe be interesting to add?
> 

Adding the scaler's name (trigger name) to the prometheus metrics seems to be an issue. It looks like the scaler's name is not being stored in the scaler itself or known outside of `scaleHandler.GetScalers(scaledObject)`, although it is being passed to the factory method - https://github.com/kedacore/keda/blob/5e6ecb33b6b16c9566d42aa412c4adbd3182bddf/pkg/scaling/scale_handler.go#L365.

Would it make sense to add additional functions to the Scaler interface to allow this data to be retrieved (ie `GetTriggerName` and `GetTriggerType`) from what is being passed in?  This obviously would require modifying all existing Scaler factory methods.

tomkerkhove commented:
My suggestion was actually for the scaler name and not the trigger name to be able to pinpoint issues with certain integrations.

On a per trigger level it would maybe be another metric or you'll have a ton of labels, but might be me. This feels more like a follow-up metric that we can still add later on.

But I'm just thinking out loud. in the end, people will ask for what they need so we get better understanding of the needs.

tbickford commented:
> My suggestion was actually for the scaler name and not the trigger name to be able to pinpoint issues with certain integrations.
> 
Can you clarify the scaler name? Is this getting the type from the scaler that is returned (ie `*scalers.rabbitMQScaler`) or something like `rabbitmq` which is represented as the triggerType?

tomkerkhove commented:
> > My suggestion was actually for the scaler name and not the trigger name to be able to pinpoint issues with certain integrations.
> 
> Can you clarify the scaler name? Is this getting the type from the scaler that is returned (ie `*scalers.rabbitMQScaler`) or something like `rabbitmq` which is represented as the triggerType?

Just the type of scaler for example rabbit MQ, Azure Service Bus, etc which we list on https://keda.sh/docs/latest/scalers/

tomkerkhove commented:
Does it have to have the exact same text? No, but preferably not with scaler in the label value

tbickford commented:
Added documentation in https://github.com/kedacore/keda-docs/pull/217

I have changed the metric for getting value to more expansive so each scaler is unique to it's trigger (when there are multiple triggers for the same scaler). Example:

```
keda_metrics_adapter_scaler_metrics_value{metric="queueLength",namespace="default",scaledObject="rabbitmq-keda-consumer-test",scaler="rabbitMQScaler",scalerIndex="0"} 0
keda_metrics_adapter_scaler_metrics_value{metric="queueLength",namespace="default",scaledObject="rabbitmq-keda-consumer-test",scaler="rabbitMQScaler",scalerIndex="1"} 8
```

tomkerkhove commented on cmd/adapter/main.go:
```diff
@@ -69,7 +71,10 @@ func (a *Adapter) makeProviderOrDie() provider.MetricsProvider {
 
-	return kedaprovider.NewProvider(logger, handler, kubeclient, namespace)
+	metricsServer := prommetrics.NewPrometheusMetricsServer()
+	go func() { metricsServer.StartServer(":9022", "/metrics") }()
```
Question - Should we make port + path configurable with this as default?

tomkerkhove replied:
Ok looks like the port is already configurable based on the docs?

> The KEDA Metrics Adapter exposes Prometheus metrics which can be scraped on port `9022` (this can be changed by setting the `metrics-port` argument for the Metrics Adapter) at `/metrics`.

So we have to update our Helm chart with this or does it already exist?

tbickford replied:
> Question - Should we make port + path configurable with this as default?

I'll add in the path as a configurable

tbickford replied:
> Ok looks like the port is already configurable based on the docs?
> 
> > The KEDA Metrics Adapter exposes Prometheus metrics which can be scraped on port `9022` (this can be changed by setting the `metrics-port` argument for the Metrics Adapter) at `/metrics`.
> 
> So we have to update our Helm chart with this or does it already exist?

The Helm Chart could be updated to specify these. I was planning on updating the Helm Chart to add the ability to either add the Prometheus scrape annotation or the Prometheus Operator Pod Monitor resource for the Metrics Adapter (this is quite common in other Helm Charts with Prometheus Metrics)

zroubalik commented:
This is an excellent contribution, thanks @tbickford!

tomkerkhove commented:
These docs on https://keda.sh/docs/2.0/operate/ are great and straightforward!

An idea that comes up today is should we:
1. Document the labels we provide?
1. Add a sample of the output to give an idea?
